pygpu/array_tools.py

```diff
@@ -0,0 +1,47 @@
+from string import Template
+from .gpuarray import GpuArray, GpuKernel
+
+
+def _generate_kernel(ctx, cols, upper=True):
+    tmpl = Template("""
+    KERNEL void extract_tri(ga_float *a, ga_uint N) {
+        unsigned int idx = blockIdx.y*blockDim.x*gridDim.x+
+                           blockIdx.x*blockDim.x+threadIdx.x;
+        unsigned int ix = idx/${cols};
+        unsigned int iy = idx%${cols};
+        if (idx < N) {
+            if (ix ${le} iy)
+                a[idx] = 0.0;
+        }
+    }
+    """)
+    if upper:
+        le = '>'
+    else:
+        le = '<'
+    src = tmpl.substitute(cols=cols, le=le)
+    spec = [GpuArray, 'uint32']
+    k = GpuKernel(src, "extract_tri", spec, context=ctx)
+    return k
+
+
+def triu(A, ctx, inplace=True):
+    if not inplace:
+        A = A.copy()
+    upper = True
+    if A.flags['F_CONTIGUOUS']:
+        upper = False
+    k = _generate_kernel(ctx, A.shape[0], upper)
```

Comment:
Based on your kernel code, the 2nd parameter of `_generate_kernel` should be the number of elements on the first dimension of the array (generally called the "Leading Dimension'), i.e number of rows for a C-array, and number of columns for a F-array. So, it is not always `A.shape[0]`. It currently works because you test with square matrices, but if you test on non-square matrices, some errors will occur.

Same remarks for `tril`.